### DISCUSSION TITLE
[Feature Request]: document and auto generate error code specs

### DISCUSSION THREAD

**Discussion by jcstein** [celestiaorg/celestia-node#3237] [Ideas]
### Implementation ideas

Astria team asked if the error conditions, along with error codes and messages, for RPCs are specified anywhere. As an example, what are the error conditions for `blob.Submit` and `blob.GetAll`?

This feature would implement specification generation for API methods.

**Comment by jcstein** [comment]
from Janis at Astria:

>> - For `blob.GetAll` we mainly want to distinguish a server or API error from "there are no blobs for the namespace you asked for". Right now we are checking if the JSONRPC response's message field contains `"blob: not found"`
>> - For `blob.Submit` it's more a reporting thing - it would be great to be specifically able to flag things like
>>    - lack of funds
>>    - oversized arguments to `blob.Submit`
> 
> Speficically for `blob.Submit` message I can think of this scenario:
> 1. astria tries submitting 2 sequencer blocks (spread across several blobs)
> 2. the total size of `[blobs]` is smaller than but still in the vicinity of 2MiB
> 3. `blob.Submit` gets rejected because many users try to get their blobs in
> 4. astria tries submitting the 2 sequencer blocks as 2 separate `blob.Submit` messages

**Comment by jcstein** [comment]
should this just be an issue? why did it get siloed to a discussion?

**Comment by liamsi** [comment]
I think:

1. we must move away from using go's error handling as an excuse to simply put error strings into the response 
2. The canonical API should specify error codes for actual errors
